ddtrace/profiling/collector/_lock.py

```diff
@@ -233,6 +235,47 @@ def __enter__(self, *args, **kwargs):
     def __exit__(self, *args, **kwargs):
         self._release(self.__wrapped__.__exit__, *args, **kwargs)
 
+    # Get lock acquire/release call location and variable name the lock is assigned to
+    def _get_lock_call_loc_with_name(self) -> typing.Optional[str]:
+        try:
+            # We expect the call stack to be like this:
+            # 0: this
+            # 1: _acquire/_release
+            # 2: acquire/release (or __enter__/__exit__)
+            # 3: caller frame
+            # And we expect addtioinal frame if WRAPT_C_EXT is False
+            frame = sys._getframe(3 if WRAPT_C_EXT else 4)
+            code = frame.f_code
+            call_loc = "%s:%d" % (os.path.basename(code.co_filename), frame.f_lineno)
+
+            var_name = None
+
+            # Search for global variables
+            for name, value in frame.f_globals.items():
+                if value == self:
+                    var_name = name
+                    break
+                for attribute in dir(value):
+                    if not attribute.startswith("__") and getattr(value, attribute) == self:
+                        var_name = attribute
+                        break
+            # Search for local variables
+            for name, value in frame.f_locals.items():
```

Comment:
All of these lookups--the list of locals, globals, and their attributes--have highly variable search bounds.

We know from some of our other reports that whenever we have instrumentation with this form, there's going to be some user who runs headfirst into the worst case scenario and then has a bad time.

Since this code will run on-GIL, assuming control from the user whenever a lock is created (or even acquired/released if my other comment isn't appropriate), I think there's value in setting a maximum search complexity.  I'd propose just counting the number of comparisons and stopping once a user-configurable maximum has been reached (it can be a profiling-related configurable setting in `ddtrace/settings/profiling.py`)